In [1]:
#import necessary packages
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import sklearn.metrics
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn import neighbors
from sklearn.preprocessing import LabelEncoder

In [2]:
#read dataset
df = pd.read_csv('AdultUCI.csv')

In [3]:
df = df.dropna()
df.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [4]:
#checking variable levels & dropping > 32
df.nunique()
df = df.drop('native-country', axis=1)
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week income  
0          2174             0              40  small  
1             0             0              13  small  
2             0             0              40  small  
3             0             0              40  small  
4             0             0              40  small

In [5]:
#encoding categorical data
label_encoders = {}
for column in df.select_dtypes(include=['object']).columns:
    label_encoder = LabelEncoder()
    df[column] = label_encoder.fit_transform(df[column])
    label_encoders[column] = label_encoder

In [6]:
#splitting dataframes to target & predictors
predictors = df.drop('income', axis=1)
target = df['income']

In [7]:
#train test split 80/20
pred_train, pred_test, tar_train, tar_test  =  train_test_split(predictors, target, test_size=.2, random_state=1)

In [9]:
#build naives bayesian on training set
bayes_classifier = GaussianNB()
gnb = bayes_classifier.fit(pred_train,tar_train)
gnb.predictions=gnb.predict(pred_test)

In [10]:
#naive bayesian confusion matrix
sklearn.metrics.confusion_matrix(tar_test,gnb.predictions)

array([[ 483, 1067],
       [ 220, 4263]])

In [11]:
#naive bayesian accuracy
bayes_accuracy = sklearn.metrics.accuracy_score(tar_test, gnb.predictions)
bayes_accuracy

0.7866732968672302

In [12]:
#build knn on training set
knn_classifier = neighbors.KNeighborsClassifier(n_neighbors=5, weights='uniform')
knn = knn_classifier.fit(pred_train,tar_train)
knn.predictions = knn.predict(pred_test)

In [13]:
#knn confusion matrix
sklearn.metrics.confusion_matrix(tar_test,knn.predictions)

array([[ 483, 1067],
       [ 332, 4151]])

In [14]:
#knn accuracy
knn_accuracy = sklearn.metrics.accuracy_score(tar_test, knn.predictions)
knn_accuracy

0.7681087352892425

**Accuracy Comparisons**

Naive Bayesian = 78.667%

K-Nearest Neighbors = 76.811%

Decision Tree = 80.457%

Random Forest = 85.049%

Best Accuracy (Descending)

1) Random Forest 2) Decision Tree 3) Naive Bayesian 4) K-Nearest Neighbors